v1.23 blog post: Pod Security Admission Graduates to Beta

jimangel opened the pull request:
Draft blog post on behalf of SIG Auth for the v1.23 release around Pod Security Admission ([supporting KEP](https://github.com/kubernetes/enhancements/issues/2579)).

~~I also had an issue with my demo until I changed `runtimeClassNames` to `runtimeClasses`. Potentially someone can weigh in on that during tech review.~~

Next Steps:

- [x] Update blog & official docs URLs (in content and at bottom)
- [x] Update ~~`kindest/node:v1.23.0` to the correct `sha256` tag~~ (using `latest` for now)
- [x] Identify the "official" blog release date (update filename / front matter)
- [x] SIG Docs review
- [x] SIG Auth tech review

Preview quick link: https://deploy-preview-30502--kubernetes-io-main-staging.netlify.app/blog/2021/12/09/pod-security-admission-beta/

/hold (until all steps are complete)
/cc @lachie83 @reylejano @liggitt

liggitt commented:
runtimeClassNames typo fixed in https://github.com/kubernetes/website/pull/30274/files

jimangel commented:
> runtimeClassNames typo fixed in https://github.com/kubernetes/website/pull/30274/files

Perfect! As long as it's technically correct, it will have no impact to the release process (and fixes the inflight v1.23 docs code section too).

jlbutler commented:
Should this be filed against main rather than the dev branch? 

/cc @karenhchu

bridgetkromhout commented on content/en/blog/_posts/2021-12-15-pod-security-admission-psa-beta.md:
```diff
@@ -0,0 +1,620 @@
+
+**Authors:** Jim Angel (Google), Lachlan Evenson (Microsoft)
+
+_Editor's note: [Jim Angel](https://twitter.com/JimmAngel) and [Lachlan Evenson](https://twitter.com/LachlanEvenson) are both [CNCF Cloud Ambassadors](https://www.cncf.io/people/ambassadors/) and wear many hats in the open source Kubernetes community. Jim is a Cloud Consultant at Google and Lachlan is a Principal Program Manager on the Azure Core Upstream._
```
>  Lachlan is a Principal Program Manager on the Azure Core Upstream.

either "Lachlan is a Principal Program Manager on the Azure Core Upstream team." or "Lachlan is a Principal Program Manager on Azure Core Upstream." would be better.

sftim replied:
I'd shift this paragraph to the foot of the article. We don't have a strong convention on this but I want readers to jump straight to the meat of the article.

lachie83 replied:
```suggestion
_Editor's note: [Jim Angel](https://twitter.com/JimmAngel) and [Lachlan Evenson](https://twitter.com/LachlanEvenson) are both [CNCF Cloud Ambassadors](https://www.cncf.io/people/ambassadors/) and wear many hats in the open source Kubernetes community. Jim is a Cloud Consultant at Google and Lachlan is a Principal Program Manager on the Azure Core Upstream team._
```

jimangel replied:
Accidentally closed this one out.

The editors note blurb was used to establish credibility before continuing to read (as seen in similar blogs). I'd rather remove it entirely over moving it to the bottom.

Thoughts?

sftim replied:
It's the Kubernetes blog and we're discussing a feature graduation. _To me_ that provides all the gravitas we need. Omitting any note like this is a form of [countersignalling](https://en.wikipedia.org/wiki/Countersignaling).

bridgetkromhout commented on content/en/blog/_posts/2021-12-15-pod-security-admission-psa-beta.md:
```diff
@@ -0,0 +1,620 @@
+   * `pod-security.kubernetes.io/<MODE>: <LEVEL>` (required)
+   * `pod-security.kubernetes.io/<MODE>-version: <VERSION>` (*optional*, defaults to latest)
+
+A specific version can be supplied for each enforcement mode. The version pins the policy to the version that was defined at that Kubernetes version. This is to increase flexibility for future  Kubernetes enhancements and unknown security issues.
```
When I read "unknown security issues" I think, "Wait a minute -

lachie83 replied:
```suggestion
A specific version can be supplied for each enforcement mode. The version pins the policy to the version that was shipped as part of the Kubernetes release. Pinning to a specific Kubernetes version allows for deterministic policy behavior while allowing flexibility for future updates to Security Policy Standards.
```

bridgetkromhout commented on content/en/blog/_posts/2021-12-15-pod-security-admission-psa-beta.md:
```diff
@@ -0,0 +1,620 @@
+
+When to `warn`?
+
+Don't use`warn` for the exact same level+version of the policy as `enforce`. In the admission sequence, if `enforce` fails, the entire sequence fails before evaluating the `warn`.
```
> use`warn`

needs a space.

tallclair replied:
There actually is a use case for this - `enforce` isn't applied to workload resources (e.g. `deployment`), so an identical `warn` label can be used to give early warning on those. I'd actually recommend this pattern in most cases, unless you're using mutating admission (which is why it's not the default... but maybe it should be?).

bridgetkromhout commented on content/en/blog/_posts/2021-12-15-pod-security-admission-psa-beta.md:
```diff
@@ -0,0 +1,620 @@
+
+**Baseline level and workload**
+
+The baseline pod demonstrates sensible defaults but does not allow privilege escalation, so the pod above would fail here too ([more info](https://kubernetes.io/docs/concepts/security/pod-security-standards/#baseline)). Baseline accepts null / no values for unset keys.
```
"more info" is not ideal link text because of how SEO and screenreaders work. I suggest moving the link to the word "baseline" at the start of the sentence, and removing the "more info".

bridgetkromhout commented on content/en/blog/_posts/2021-12-15-pod-security-admission-psa-beta.md:
```diff
@@ -0,0 +1,620 @@
+
+**Restricted level and workload**
+
+Restricted requires rejection of all privileged parameters. It is the most secure with a trade-off for complexity ([more info](https://kubernetes.io/docs/concepts/security/pod-security-standards/#restricted)). The restricted policy allows containers to add the `NET_BIND_SERVICE` capability.
```
Same comment as earlier - I suggest that you make the word "Restricted" the link text, not "more info".

sftim commented on content/en/blog/_posts/2021-12-15-pod-security-admission-psa-beta.md:
```diff
@@ -0,0 +1,620 @@
+layout: blog
+title: 'Pod Security Admission (PSA) Graduates to Beta'
+date: 2021-12-15
+slug: pod-security-admission-psa-beta
```
Avoiding a kind of tautology:
```suggestion
slug: pod-security-admission-beta
```

sftim commented on content/en/blog/_posts/2021-12-15-pod-security-admission-psa-beta.md:
```diff
@@ -0,0 +1,620 @@
+
+## Why Pod Security Admission
+
+Pod Security Admission overcomes the shortcomings of, deprecated, Pod Security Policy (PSP):
```
I'd write:
```suggestion
Pod Security Admission overcomes key shortcomings of Kubernetes' existing, but deprecated,
PodSecurityPolicy (PSP) mechanism:
```
Is that better?

lachie83 replied:
Yep. I like it

sftim commented on content/en/blog/_posts/2021-12-15-pod-security-admission-psa-beta.md:
```diff
@@ -0,0 +1,620 @@
+Pod Security Admission overcomes the shortcomings of, deprecated, Pod Security Policy (PSP):
+
+   * Policy authorization model — challenging to deploy with controllers.
+   * Challenging to enable — no dry-run/audit capabilities made it hard to enable.
```
```suggestion
   * Risks around switching — a lack of dry-run/audit capabilities made it hard to enable PodSecurityPolicy 
     safely in your cluster.
```
?

sftim commented on content/en/blog/_posts/2021-12-15-pod-security-admission-psa-beta.md:
```diff
@@ -0,0 +1,620 @@
+
+The shortcomings of PSP made it very difficult to use which led the community to reevaluate whether or not a better implementation could achieve the same goals. One of those goals was to provide an out-of-the-box solution to apply security best practices. Pod Security Admission ships with predefined Pod Security levels that a cluster administrator can configure to meet the desired security postur ...
+
+It's important to note that Pod Security Admission doesn't have complete feature parity with the deprecated Pod Security Policy. Specifically, it doesn't have the ability to `mutate` or change Kubernetes resources to auto-remediate a policy violation on behalf of the user. Additionally, it doesn't provide fine-grain control over each allowed field and value within a pod specification or any other ...
```
Should we also mention there are other projects with more fine grained control? https://kubernetes.io/docs/concepts/security/pod-security-standards/#faq has our official list of alternatives.

sftim commented on content/en/blog/_posts/2021-12-15-pod-security-admission-psa-beta.md:
```diff
@@ -0,0 +1,620 @@
+
+In addition to modes you can also pin the policy to a specific version for example v1.22. Pinning to a specific version allows the behavior to remain consistent as policy changes happen over Kubernetes releases.
+
+## Hands on demo
```
Can instead
- link to https://kubernetes.io/docs/tasks/configure-pod-container/enforce-standards-admission-controller/ instead?
- copy the steps from there?

_or_ merge these steps into that task page?

I don't want us to have two pages that document different ways to try the same thing, on release day. That'd feel counter-productive as readers might wonder which they should follow.

sftim replied:
You could also add this content within https://docs.k8s.io/tutorials/ - I think that'd be a really good choice.

jimangel replied:
@sftim do you think the additional content should be included with the blog PR or a follow-up? Happy to open a tracking issue. Also, are you saying to link the entire hands-on demo portion to the official docs and move the demo there?

I like both ideas, if I understand them correctly, but it might be worth tracking vs. implementing.

sftim commented on content/en/blog/_posts/2021-12-15-pod-security-admission-psa-beta.md:
````diff
@@ -0,0 +1,620 @@
+
+The output is similar to this:
+
+```shell
````
This isn't a shell script.

sftim commented on content/en/blog/_posts/2021-12-15-pod-security-admission-psa-beta.md:
````diff
@@ -0,0 +1,620 @@
+Error from server (Forbidden): error when creating "STDIN": pods "nginx-privileged" is forbidden: violates PodSecurity "baseline:latest": host namespaces (hostNetwork=true)
+```
+
+#### Yes!!! It worked! 🎉🎉🎉
````
nit (simpler fragment identifier):
```suggestion
#### Yes!!! It worked! 🎉🎉🎉 {#it-worked}
```

sftim commented on content/en/blog/_posts/2021-12-15-pod-security-admission-psa-beta.md:
````diff
@@ -0,0 +1,620 @@
+cat /etc/kubernetes/manifests/kube-apiserver.yaml 
+```
+
+**UPDATE:** The baseline policy permits `allowPrivilegeEscalation`. While I cannot see the Pod Security default levels of enforcement, they are there. Let's try to provide a manifest that violates the baseline by requesting hostNetwork access.
+
+```shell
+cat <<EOF | kubectl -n test-defaults apply -f -
+apiVersion: v1
+kind: Pod
+metadata:
+  name: nginx-privileged
+spec:
+  containers:
+  - name: nginx
+    image: nginx
+  hostNetwork: true
+EOF
+```
````
aside: In docs, but not blog articles, we usually use the `codenew` shortcode and tell people to apply a manifest by its URL.

sftim commented on content/en/blog/_posts/2021-12-15-pod-security-admission-psa-beta.md:
````diff
@@ -0,0 +1,620 @@
+
+```shell
+# shell into the control plane
+docker exec -it kind-control-plane bash
````
BTW this assumes you're using kind with Docker (and not Podman).

sftim commented on content/en/blog/_posts/2021-12-15-pod-security-admission-psa-beta.md:
````diff
@@ -0,0 +1,620 @@
+
+Hmm... yep. Our default `warn` level is working at least.
+
+```shell
````
This isn't a shell script.

sftim commented on content/en/blog/_posts/2021-12-15-pod-security-admission-psa-beta.md:
````diff
@@ -0,0 +1,620 @@
+EOF
+```
+
+Hmm... yep. Our default `warn` level is working at least.
````
Avoid writing “our” if the sense you're using it in includes the reader; that form can be confusing (and we avoid it even more in the docs).

sftim commented on content/en/blog/_posts/2021-12-15-pod-security-admission-psa-beta.md:
````diff
@@ -0,0 +1,620 @@
+No LimitRange resource.
+```
+
+Can we deploy a workload that is privileged?
````
Avoid writing “we” if the sense you're using it in includes the reader; that form can be confusing (and we avoid it even more in the docs).

sftim commented on content/en/blog/_posts/2021-12-15-pod-security-admission-psa-beta.md:
````diff
@@ -0,0 +1,620 @@
+
+Same.
+
+```shell
````
This isn't a shell script.

sftim commented on content/en/blog/_posts/2021-12-15-pod-security-admission-psa-beta.md:
````diff
@@ -0,0 +1,620 @@
+
+The output is similar to this:
+
+```shell
````
This isn't a shell script.

sftim reviewed (commented):
Some early feedback. I realise it's a WIP PR so please see everything here as a suggestion only.

sftim commented:
We should get a tech review on this from SIG Auth (which I hope will be easy to get because of the eyes on this already)

lachie83 commented:
> We should get a tech review on this from SIG Auth (which I hope will be easy to get because of the eyes on this already)

I will reach out to @ritazh from SIG Auth to see who can help with tech review.

jimangel commented:
> Should this be filed against main rather than the dev branch?

If we know the blog release date before the Kubernetes version is released, it shouldn't matter. As it could merge and not go live until the date.

I think you're right though, historically we've opened blogs against `main` since they would get merged / go live post-release.

I'm good with either option, let me know a decision and I can act. I would love someone closer to the blog subproject to weigh in with a suggestion.

tallclair commented on content/en/blog/_posts/2021-12-15-pod-security-admission-psa-beta.md:
```diff
@@ -0,0 +1,620 @@
+### Configure Pod Security Admission
+
+Policies are applied to a namespace via labels. These labels are as follows:
+   * `pod-security.kubernetes.io/<MODE>: <LEVEL>` (required)
```
nit: I don't want people to think the are required to set this on every namespace as soon as they upgrade to v1.23. Maybe something like this instead?
```suggestion
   * `pod-security.kubernetes.io/<MODE>: <LEVEL>` (required to enable pod security)
```

tallclair commented on content/en/blog/_posts/2021-12-15-pod-security-admission-psa-beta.md:
```diff
@@ -0,0 +1,620 @@
+
+## Introduction
+
+With the release of Kubernetes v1.23, [Pod Security Admission (PSA)](https://kubernetes.io/docs/concepts/security/pod-security-admission/) has now entered beta. PSA is a [built-in](https://kubernetes.io/docs/reference/access-authn-authz/admission-controllers/) admission controller that evaluates pod specifications against a predefined set of [Pod Security Standards](https://kubernetes.io/docs/con ...
```
nit: I prefer to add line breaks after ~100 characters or so for ease of review. It makes it easier to comment on something in the middle of a paragraph, and prevents multiple comment threads from being collapsed into one.

sftim replied:
In docs, I'd call the new feature Pod Security Admission or Pod Security admission. There's no resource in the API that you can fetch with a kind called PodSecurity, and UpperCamelCase names give readers a hint that they would see that.

Separately, line breaks make articles easier to localize (some localizations pick up blog articles).

tallclair commented on content/en/blog/_posts/2021-12-15-pod-security-admission-psa-beta.md:
```diff
@@ -0,0 +1,620 @@
+
+A specific version can be supplied for each enforcement mode. The version pins the policy to the version that was defined at that Kubernetes version. This is to increase flexibility for future  Kubernetes enhancements and unknown security issues.
+
+From earlier, the modes are:
+   * `enforce` — Any Pods that violate the policy will be rejected
+   * `audit` — Pods with violations will be allowed and an audit annotation will be added
+   * `warn` — Pods that violate the policy will be allowed and a warning message will be sent back to the user.
```
Rather than repeating this whole thing, maybe just enumerate the modes? 
```suggestion
From earlier, recall that the possible modes are `enforce`, `audit` and `warn`.
```

tallclair commented on content/en/blog/_posts/2021-12-15-pod-security-admission-psa-beta.md:
```diff
@@ -0,0 +1,620 @@
+   * `audit` — Pods with violations will be allowed and an audit annotation will be added
+   * `warn` — Pods that violate the policy will be allowed and a warning message will be sent back to the user.
+
+> Note: The default for enforce/warn/audit is `privileged` if not specified
```
I know there's a special directive for notes on the docs pages. I don't know whether it applies to blog posts...

sftim replied:
I wouldn't use those callouts on blog articles, but I also _wouldn't_ use a quote (block quote, here) to simulate them.
Just use running text, and bold / italics if you feel the need for it.

jimangel replied:
Removed entire "Note" as it doesn't add any value to the reader's experience.

tallclair commented on content/en/blog/_posts/2021-12-15-pod-security-admission-psa-beta.md:
```diff
@@ -0,0 +1,620 @@
+spec:
+  containers:
+  - name: nginx
+    image: nginx
```
nit: I'd prefer to use the pause image (`k8s.gcr.io/pause`) in examples that don't need to actually run anything.

It's less likely to terminate for other reasons, and it's an image we own (not promoting other projects)

sftim replied:
Maybe add a comment about why we picked it?

tallclair commented on content/en/blog/_posts/2021-12-15-pod-security-admission-psa-beta.md:
````diff
@@ -0,0 +1,620 @@
+kubectl label --overwrite ns kube-test-restricted \
+  pod-security.kubernetes.io/enforce=restricted \
+  pod-security.kubernetes.io/audit=restricted
+```
````
Consider creating the namespace before the example that uses them (or even reusing & relabeling a single namespace). That way the reader doesn't need to check back here to remember how each namespace was configured.

tallclair commented on content/en/blog/_posts/2021-12-15-pod-security-admission-psa-beta.md:
````diff
@@ -0,0 +1,620 @@
+EOF
+```
+
+The output is similar to this:
````
it's easy to miss that `pod/nginx-baseline created`. Consider highlighting it here.
```suggestion
The output is similar to the following. Note that the warnings match the error message from the test above, but the pod is still successfully created.
```

tallclair commented on content/en/blog/_posts/2021-12-15-pod-security-admission-psa-beta.md:
```diff
@@ -0,0 +1,620 @@
+<!-- TODO: link to the official docs after posted: https://github.com/kubernetes/website/blob/dev-1.23/content/en/docs/tasks/configure-pod-container/enforce-standards-admission-controller.md?plain=1#L25-L55
+--->
+
+<!-- TODO: is this a BUG?:
+It errors with runtimeClassNames: [] and runs without it, tested on both alphav1/betav1 apis.
```
Looks like you already corrected the fieldname to `runtimeClasses` (I think we had mis-documented it)

tallclair commented on content/en/blog/_posts/2021-12-15-pod-security-admission-psa-beta.md:
````diff
@@ -0,0 +1,620 @@
+
+Run the following command:
+
+```
+kubectl get --raw /metrics | grep pod_security_evaluations_total
+```
````
Oh, right, nice. This is a better way to check if it's working/enabled (section above).

jimangel replied:
Do you think it's worth stripping the user journey section of demo and consolidating it by starting with the `raw metrics` check to validate followed by a clean(er) working demo?

I like both approaches. I do enjoy watching "the journey" but other folks might want just raw / working content.

reylejano replied:
Since this is a blog post, I like the user journey section (imo)

tallclair reviewed (commented):
Awesome, thanks for writing this!

jimangel commented on content/en/blog/_posts/2021-12-15-pod-security-admission-psa-beta.md:
```diff
@@ -0,0 +1,620 @@
+
+It's important to note that Pod Security Admission doesn't have complete feature parity with the deprecated Pod Security Policy. Specifically, it doesn't have the ability to `mutate` or change Kubernetes resources to auto-remediate a policy violation on behalf of the user. Additionally, it doesn't provide fine-grain control over each allowed field and value within a pod specification or any other ...
+
+Pod Security Admission also adheres to Kubernetes best practices of declarative object management by denying resources that violate the policy. This requires resources to be updated in source repositories, and tooling to be updated prior to being deployed to Kubernetes.
```
If you're referring to "Pod Security Admission also adheres to Kubernetes best practices of declarative object management by denying resources that violate the policy. This requires resources to be updated in source repositories, and tooling to be updated prior to being deployed to Kubernetes."

The intention is to call out the explicit declarative nature of Pod Security (and any admission controller), whereas PSP had the ability to introspect CRI information to make decisions. For example, if you built a container as non-root and didn't include a `runAsUser` would pass PSP but fail Pod Security (circumstances depending).

Open to alternative ways to draw attention to this paradigm shift from PSP to "normal" admission behavior.

sftim commented:
>  If we know the blog release date before the Kubernetes version is released, it shouldn't matter. As it could merge and not go live until the date.
> I think you're right though, historically we've opened blogs against main since they would get merged / go live post-release.

I would have this ready-to-merge but not merge it until the release is confirmed. And I'd target _main_.
In essence, we shouldn't merge any announcement for publication until we are **certain** of the publication date.

sftim commented on content/en/blog/_posts/2021-12-15-pod-security-admission-psa-beta.md:
```diff
@@ -0,0 +1,620 @@
+
+### Confirm PSA is enabled
+
+The best way to [confirm the API's default enabled plugins](https://kubernetes.io/docs/reference/access-authn-authz/admission-controllers/#which-plugins-are-enabled-by-default) is to check the Kubernetes API container's help arguments.
```
IMO we should document this _somewhere_. Folks will want to know!

sftim commented:
I haven't tested the steps. The text is already looking great.

ritazh commented on content/en/blog/_posts/2021-12-15-pod-security-admission-beta.md:
```diff
@@ -0,0 +1,612 @@
+
+**Authors:** Jim Angel (Google), Lachlan Evenson (Microsoft)
+
+With the release of Kubernetes v1.23, [Pod Security Admission (PSA)](/docs/concepts/security/pod-security-admission/) has now entered beta. PSA is a [built-in](/docs/reference/access-authn-authz/admission-controllers/) admission controller that evaluates pod specifications against a predefined set of [Pod Security Standards](/docs/concepts/security/pod-security-standards/) and determines whether  ...
```
nit: 
> will be removed in v1.25

jimangel replied:
Good callout! I wonder if we should instead update `which was deprecated in v1.21` to `which is deprecated in v1.21` instead, to keep everything present tense (inline with the [style guide](https://kubernetes.io/docs/contribute/style/style-guide/#use-present-tense)).

However, not sure how strict that rule would be in this context (speaking as a blog, in a point of time, where one item is past tense and one is future tense) cc @sftim

sftim replied:
“_which was deprecated in v1.21_” is how I'd say it conversationally:
- A: “I'm using PodSecurityPolicy”
- B: “Kubernetes deprecated that in v1.21”
- A: “It was deprecated in v1.21? I'd better go and read about alternatives!”
- B: “Here's my rate card, let me know if you'd like help here.”

jimangel replied:
With that same mindset, you think we should use `s/is removed in v1.25/will be removed in v1.25`?

ritazh commented on content/en/blog/_posts/2021-12-15-pod-security-admission-beta.md:
```diff
@@ -0,0 +1,612 @@
+
+The shortcomings of PSP made it very difficult to use which led the community to reevaluate whether or not a better implementation could achieve the same goals. One of those goals was to provide an out-of-the-box solution to apply security best practices. Pod Security Admission ships with predefined Pod Security levels that a cluster administrator can configure to meet the desired security postur ...
+
+It's important to note that Pod Security Admission doesn't have complete feature parity with the deprecated PodSecurityPolicy. Specifically, it doesn't have the ability to `mutate` or change Kubernetes resources to auto-remediate a policy violation on behalf of the user. Additionally, it doesn't provide fine-grain control over each allowed field and value within a pod specification or any other K ...
```
should we link to https://github.com/open-policy-agent/gatekeeper-library instead since that is where we maintain the community-owned policies?

lachie83 replied:
Because we aren't mentioning that there are PodSecurityPolicy/PodSecurity equivalent policies that Gatekeeper implements I don't think it's worth digging too deep in this blog and hope that if the reader is interested the Gatekeeper docs would eventually point them at the policy library.

ritazh commented on content/en/blog/_posts/2021-12-15-pod-security-admission-beta.md:
```diff
@@ -0,0 +1,612 @@
+   * `audit` — Pods with violations will be allowed and an audit annotation will be added
+   * `warn` — Pods that violate the policy will be allowed and a warning message will be sent back to the user.
+
+In addition to modes you can also pin the policy to a specific version for example v1.22. Pinning to a specific version allows the behavior to remain consistent as policy changes happen over Kubernetes releases.
```
should we add something about EOL and support for older k8s release policies?

ritazh commented on content/en/blog/_posts/2021-12-15-pod-security-admission-beta.md:
```diff
@@ -0,0 +1,612 @@
+   * `pod-security.kubernetes.io/<MODE>: <LEVEL>` (required)
+   * `pod-security.kubernetes.io/<MODE>-version: <VERSION>` (*optional*, defaults to latest)
+
+A specific version can be supplied for each enforcement mode. The version pins the policy to the version that was shipped as part of the Kubernetes release. Pinning to a specific Kubernetes version allows for deterministic policy behavior while allowing flexibility for future updates to Security Policy Standards
```
nit:
> future updates to Pod Security Standards

ritazh commented on content/en/blog/_posts/2021-12-15-pod-security-admission-beta.md:
````diff
@@ -0,0 +1,612 @@
+The output is similar to this:
+
+```
+# HELP pod_security_evaluations_total [ALPHA] Number of policy evaluations that occurred, not counting ignored or exempt requests.
````
Is this alpha specifically for the `pod_security_evaluations_total ` metric?

jimangel replied:
cc @liggitt who shared the `metrics` pro-tip with me.

liggitt replied:
yes, the metric is alpha (metrics currently only have ALPHA or STABLE stability levels)

ritazh commented on content/en/blog/_posts/2021-12-15-pod-security-admission-beta.md:
````diff
@@ -0,0 +1,612 @@
+No LimitRange resource.
+```
+
+Let's create a new namespace and see if the labels apply there.
````
might be good to add a snippet about what happens when the label is removed from the namespace and how that impacts update operations for existing pods

jimangel replied:
Any suggestions or recommendations? I like the idea; not sure the best way to introduce it without feeling forced / distracting.

FlorentATo commented on content/en/blog/_posts/2021-12-15-pod-security-admission-beta.md:
```diff
@@ -0,0 +1,635 @@
+Policies are applied in a specific mode. Multiple modes (with different policies) can be set on the same namespace. Here is a list of modes:
+   * `enforce` — Any Pods that violate the policy will be rejected
+   * `audit` — Violations will be recorded as an annotation in the audit logs, but don't affect whether the pod is allowed.
+   * `warn` — Violations will send a warning message back to the user, but don't affect whether the pod is allowed.sent back to the user.
```
```suggestion
   * `warn` — Violations will send a warning message back to the user, but don't affect whether the pod is allowed.
```

FlorentATo commented on content/en/blog/_posts/2021-12-15-pod-security-admission-beta.md:
````diff
@@ -0,0 +1,635 @@
+kind create cluster --image kindest/node:v1.23.0
+```
+
+It might take awhile to start and once it's started it might take a minute or so before the node becomes ready.
````
```suggestion
It might take a while to start and once it's started it might take a minute or so before the node becomes ready.
```

FlorentATo commented on content/en/blog/_posts/2021-12-15-pod-security-admission-beta.md:
```diff
@@ -0,0 +1,635 @@
+   * `pod-security.kubernetes.io/<MODE>: <LEVEL>` (required to enable pod security)
+   * `pod-security.kubernetes.io/<MODE>-version: <VERSION>` (*optional*, defaults to latest)
+
+A specific version can be supplied for each enforcement mode. The version pins the policy to the version that was shipped as part of the Kubernetes release. Pinning to a specific Kubernetes version allows for deterministic policy behavior while allowing flexibility for future updates to Security Policy Standards From earlier, recall that the possible modes are `enforce`, `audit` and `warn`.
```
The last part seems disconnected from the paragraph.

FlorentATo commented on content/en/blog/_posts/2021-12-15-pod-security-admission-beta.md:
```diff
@@ -0,0 +1,635 @@
+
+A specific version can be supplied for each enforcement mode. The version pins the policy to the version that was shipped as part of the Kubernetes release. Pinning to a specific Kubernetes version allows for deterministic policy behavior while allowing flexibility for future updates to Security Policy Standards From earlier, recall that the possible modes are `enforce`, `audit` and `warn`.
+
+When to use `warn`?
+
+Don't use `warn` for the exact same level+version of the policy as `enforce`. In the admission sequence, if `enforce` fails, the entire sequence fails before evaluating the `warn`. The typical uses for `warn` are:
```
```suggestion
When _not_ to use `warn`?

Don't use `warn` for the exact same level+version of the policy as `enforce`. In the admission sequence, if `enforce` fails, the entire sequence fails before evaluating the `warn`. 

When to use `warn`?
The typical uses for `warn` are:
```

FlorentATo commented on content/en/blog/_posts/2021-12-15-pod-security-admission-beta.md:
```diff
@@ -0,0 +1,635 @@
+
+**Privileged level and workload**
+
+For the privileged pod, allow [privilege escalation](/docs/concepts/security/pod-security-standards/#privileged). This allows the process inside a container to gain new processes and can be dangerous if untrusted.
```
```suggestion
For the privileged pod, use the [privileged policy](/docs/concepts/security/pod-security-standards/#privileged). This allows the process inside a container to gain new processes (aka privilege escalation) and can be dangerous if untrusted.
```

FlorentATo commented on content/en/blog/_posts/2021-12-15-pod-security-admission-beta.md:
````diff
@@ -0,0 +1,635 @@
+  pod-security.kubernetes.io/audit=restricted
+```
+
+Next, try to deploy a provileged workload in the namespace.
````
```suggestion
Next, try to deploy a privileged workload in the namespace.
```

FlorentATo commented on content/en/blog/_posts/2021-12-15-pod-security-admission-beta.md:
```diff
@@ -0,0 +1,635 @@
+
+**Baseline level and workload**
+
+The [baseline](/docs/concepts/security/pod-security-standards/#baseline) pod demonstrates sensible defaults but does not allow privilege escalation, so the pod above would fail here too. Baseline accepts null / no values for unset keys.
```
```suggestion
The [baseline policy](/docs/concepts/security/pod-security-standards/#baseline) demonstrates sensible defaults but does not allow privilege escalation, so the pod above would fail here too. Baseline accepts null / no values for unset keys.
```

FlorentATo commented on content/en/blog/_posts/2021-12-15-pod-security-admission-beta.md:
```diff
@@ -0,0 +1,635 @@
+
+**Restricted level and workload**
+
+[Restricted](/docs/concepts/security/pod-security-standards/#restricted) requires rejection of all privileged parameters. It is the most secure with a trade-off for complexity. The restricted policy allows containers to add the `NET_BIND_SERVICE` capability.
```
```suggestion
The [restricted policy](/docs/concepts/security/pod-security-standards/#restricted) requires rejection of all privileged parameters. It is the most secure with a trade-off for complexity. The restricted policy allows containers to add the `NET_BIND_SERVICE` capability.
```

FlorentATo commented on content/en/blog/_posts/2021-12-15-pod-security-admission-beta.md:
````diff
@@ -0,0 +1,612 @@
+EOF
+```
+
+We now have a default baseline policy, next pass it to the kind configuration to enable the `--admission-control-config-file` API server argument and pass the policy file.
````
```suggestion
We now have a default baseline policy. Next, pass it to the kind configuration to enable the `--admission-control-config-file` API server argument and pass the policy file.
```

FlorentATo commented on content/en/blog/_posts/2021-12-15-pod-security-admission-beta.md:
````diff
@@ -0,0 +1,635 @@
+    readOnly: true
+EOF
+```
+
````
```suggestion

Create a new cluster using the kind configuration file defined above.

```

jimangel commented on content/en/blog/_posts/2021-12-15-pod-security-admission-beta.md:
````diff
@@ -0,0 +1,766 @@
+
+Example `audit-policy.yaml` configuration tuned for Pod Security events:
+
+```
+apiVersion: audit.k8s.io/v1
+kind: Policy
+rules:
+- level: RequestResponse
+  resources:
+    - group: "" # core API group
+      resources: ["pods", "deployments", "statefulsets", "daemonsets"]
+  verbs: ["create"]
+  omitStages:
+    - "RequestReceived"
+    - "ResponseStarted"
+    - "Panic"
+```
````
Is there a better policy / filter to use here (for simplicity)?
cc @tallclair @liggitt

jimangel replied:
+ @kubernetes/sig-security-pr-reviews

FlorentATo commented:
Thank you @jimangel for applying my suggestions! Looking forward to use that feature live soon :)

jimangel commented on content/en/blog/_posts/2021-12-15-pod-security-admission-beta.md:
```diff
@@ -0,0 +1,766 @@
+
+A monitoring tool could ingest these metrics too for reporting, assessments, or measuring trends.
+
+### Auditing
```
```suggestion
## Auditing
```

jimangel commented on content/en/blog/_posts/2021-12-15-pod-security-admission-beta.md:
```diff
@@ -0,0 +1,766 @@
+
+If using a file (`--audit-log-path`), run `cat /PATH/TO/API-AUDIT/LOG.log | grep "is forbidden:"` to see all rejected workloads audited.
+
+# PSP migrations
```
```suggestion
## PSP migrations
```

jimangel commented:
/retitle v1.23 blog post: Pod Security Admission Graduates to Beta

PushkarJ commented:
Hi All, we discussed about this wonderful blog PR, in today's SIG Security call, so wanted to share here the summary for those who could not attend and get any other feedback you may have. 

After discussion with SIG Docs couple of weeks back, I opened a PR for Tutorials on Pod Security Admission: https://github.com/kubernetes/website/pull/30422. It has a step by step guide on how to enable `baseline` Pod Security Standards at cluster and namespace level with a working shell-script that does all of it together. The Tutorials will track Pod Security Admission feature through current and future releases and will be updated appropriately. 

On the other hand, the blog would serve as a point in time guide on how Pod Security Admission feature works for v1.23 and will mimic a user trying this new feature and playing around with it. 

In summary, it makes sense for both of these artifacts to co-exist and serve a complementary purpose and audience. Having said that both the blog and tutorial should be cross-referenced and highlighted so that either of them are not missed by folks reading or coming across only one of them. Comments / suggestions to facilitate this cross-linking and potential deduplication and consistency fixes for this blog PR and the tutorial PR are appreciated. @jimangel and I have been discussing this since last week and are both aligned on this topic!

It would also be ideal for both to be merged / published around the same dates so that they are available for consumption and most importantly the cross-links work. (looking towards @sftim on their advise on how we could do align the dates from logistics perspective)

PushkarJ commented on content/en/blog/_posts/2021-12-15-pod-security-admission-beta.md:
```diff
@@ -0,0 +1,774 @@
+  name: busybox-privileged
+spec:
+  containers:
+  - name: busybox
```
Note to myself: Tutorial is using `nginx` whereas blog is using `busybox`. Switch over the tutorial to use `busybox`

jimangel replied:
I started using `nginx` and the comment came up that we should use something community owned, like `k8s.gcr.io/pause`, but I ran into a couple snags getting the demo to work smoothly.

I switched back to `nginx`, then later ran into issues when I was applying to the restricted profile (e.g: nginx is built / runs as root, so in my demo, I could either use `nginxinc/nginx-unprivileged` (not preferred) or something less opinionated about the runAsUser so I went with `busybox` + `sleep`. I'm not committed to using `busybox` if there's a compelling reason to use something else!

PushkarJ replied:
Oh thanks for the history on why you switched to `busybox` from `nginx`. After reading that, makes more sense to continue to use `busybox`!!

PushkarJ commented on content/en/blog/_posts/2021-12-15-pod-security-admission-beta.md:
````diff
@@ -0,0 +1,774 @@
+
+Clean up the demo (restricted pod and namespace) with:
+
+```shell
````
Note to Self: Add a clean up section for tutorial

PushkarJ replied:
Tutorial now has a clean up section

PushkarJ commented on content/en/blog/_posts/2021-12-15-pod-security-admission-beta.md:
```diff
@@ -0,0 +1,774 @@
+            pathType: "DirectoryOrCreate"
+  # mount the local file on the control plane
+  extraMounts:
+  - hostPath: ./pod-security.yaml
```
Maybe use `/tmp` directory for consistency with tutorial: https://github.com/kubernetes/website/blob/004e9a48b3cb7912c85a24f1fcb6812f91abb688/content/en/docs/tutorials/pod-security/cluster-level-pss.md#configure-the-api-server and easier clean up?

jimangel replied:
IIRC `/tmp` works but, given the permissions that /tmp usually has, you're mounting a very open dir into the containers for potential abuse. While I think it's unlikely, it's probably not a great practice if someone adopts it in production. I like the idea of taking the current directory for the users interpretation.

I don't think either of the tutorials need to change but I could be sold on unifying. (maybe it's a good idea to have diverging examples?)

PushkarJ commented on content/en/blog/_posts/2021-12-15-pod-security-admission-beta.md:
````diff
@@ -0,0 +1,774 @@
+pod_security_evaluations_total{decision="deny",mode="enforce",policy_level="baseline",policy_version="latest",request_operation="create",resource="pod",subresource=""} 1
+pod_security_evaluations_total{decision="deny",mode="warn",policy_level="restricted",policy_version="latest",request_operation="create",resource="controller",subresource=""} 2
+pod_security_evaluations_total{decision="deny",mode="warn",policy_level="restricted",policy_version="latest",request_operation="create",resource="pod",subresource=""} 2
+```
````
This is my favorite section of the blog!!

PushkarJ commented on content/en/blog/_posts/2021-12-15-pod-security-admission-beta.md:
```diff
@@ -0,0 +1,774 @@
+
+## Wrap up
+
+Pod Security is a promising new feature that provides an out-of-the-box way to allow users to improve the security posture of their workloads. Like any new enhancement that has matured to beta, we ask that you try it out, provide feedback, or share your experience via either raising a Github issue or joining SIG Auth community meetings. It's our hope that Pod Security will be deployed on every cl ...
```
```suggestion
Pod Security is a promising new feature that provides an out-of-the-box way to allow users to improve the security posture of their workloads. Like any new enhancement that has matured to beta, we ask that you try it out, provide feedback, or share your experience via either raising a Github issue or joining SIG Auth community meetings. It's our hope that Pod Security will be deployed on every cluster in our ongoing pursuit as a community to make Kubernetes security a priority.
For a step by step guide on how to enable "baseline" Pod Security Standards with Pod Security Admission feature please refer to these dedicated [tutorials](docs/tutorials/pod-security) that cover the configuration needed at cluster level and namespace level. 
```

PushkarJ replied:
Felt like highlighting the tutorial as part of the wrap up would make it easier to not miss them v/s clubbing them together with other references. Open to other ideas too

PushkarJ commented on content/en/blog/_posts/2021-12-15-pod-security-admission-beta.md:
```diff
@@ -0,0 +1,774 @@
+    defaults:
+      enforce: "baseline"
+      enforce-version: "latest"
+      audit: "baseline"
```
```suggestion
      audit: "restricted"
```

jimangel replied:
I'm ok with this suggestion, but curious about your reasoning why? I don't care much if it's `baseline` or `restricted` for auditing, but I approached it as a cluster operator of, "I care historically what's allowed or rejected by PSA" vs. auditing a theoretical higher level of restrictions. I could see value in this approach though if you were interested in, "How many workloads do I allow today that would fail a more restrictive level?" but maybe I would use `warn` for that.

I see the modes like:
- `warn` == for developers / CD signal
- `audit` == for operators / compliance

No matter what approach is used, it might be worth talking about "why;" kind of similar to how we talk about when to `warn`. Wdyt?

PushkarJ replied:
To be honest, I was only trying to align the label here with the text above and the tutorial for consistency :D 

In terms of personas though, you make a good point. I was mostly thinking about the latter when selecting these levels for audit i.e. "How many workloads do I allow today that would fail a more restrictive level?". But the other scenario is valid too. 

As long as we pick one consistently and share the rationale behind it prior to the `yaml`, Happy with either of them.

jimangel replied:
I think it's worth adding a blurb about why the levels are chosen, similar to the the "when to warn?" section. I'll probably keep it as is with `baseline` though. Maybe you could do the same for using `restricted` in the tutorial? Open to alternative suggestions too!

PushkarJ replied:
yes that works!

PushkarJ commented on content/en/blog/_posts/2021-12-15-pod-security-admission-beta.md:
```diff
@@ -0,0 +1,774 @@
+
+There is no runtime configurable API for this resource and a cluster administrator would need to specify a path to the file below via the `--admission-control-config-file` flag on the API server. 
+
+In the following resource we are enforcing the baseline policy and warning and auditing the baseline policy. We are also making the kube-system namespace exempt from this policy.
```
```suggestion
In the following resource we are enforcing the baseline policy and warning and auditing the restricted policy. We are also making the kube-system namespace exempt from this policy.
```

PushkarJ reviewed (commented):
Added a few comments / suggestions and a couple of note to self for updating the tutorials

jimangel commented on content/en/blog/_posts/2021-12-15-pod-security-admission-beta.md:
````diff
@@ -0,0 +1,774 @@
+### Deploy a kind cluster
+
+```shell
+kind create cluster --image kindest/node:v1.23.0
````
I had the same issue and built one too (then updated PR not to use my local tag). I have it listed in my "steps before hold can be lifted" to include changing to the latest `sha256` tag. With the PR targeting `main` it can merge at anytime, even post-release.

liggitt commented on content/en/blog/_posts/2021-12-15-pod-security-admission-beta.md:
```diff
@@ -0,0 +1,774 @@
+
+The shortcomings of PSP made it very difficult to use which led the community to reevaluate whether or not a better implementation could achieve the same goals. One of those goals was to provide an out-of-the-box solution to apply security best practices. Pod Security ships with predefined Pod Security levels that a cluster administrator can configure to meet the desired security posture.
+
+It's important to note that Pod Security doesn't have complete feature parity with the deprecated PodSecurityPolicy. Specifically, it doesn't have the ability to `mutate` or change Kubernetes resources to auto-remediate a policy violation on behalf of the user. Additionally, it doesn't provide fine-grain control over each allowed field and value within a pod specification or any other Kubernetes  ...
```
nit: not sure why mutate is quoted with back-ticks... it's not really a keyword / code verb

liggitt replied:
nit: make fine-grain and fine-grained consistent?

liggitt replied:
we were pretty careful not to pick just one external policy project when linking to alternatives in https://kubernetes.io/docs/concepts/security/pod-security-standards/#what-s-the-difference-between-a-security-profile-and-a-security-context ... not sure if you want to be similarly noncommittal here

liggitt commented on content/en/blog/_posts/2021-12-15-pod-security-admission-beta.md:
```diff
@@ -0,0 +1,774 @@
+### Pod Security Standards
+    
+In order to use Pod Security we first need to understand [Pod Security Standards](/docs/concepts/security/pod-security-standards/). These standards define three different policy levels that range from permissive to restrictive. These levels are as follows:
+   * Privileged — open and unrestricted
+   * Baseline — Covers known privilege escalations while minimizing restrictions
+   * Restricted — Highly restricted, hardening against known and unknown privilege escalations. May cause compatibility issues
```
I might suggest making `privileged`, `baseline`, and `restricted` lower-case and code-quoted, since these are keywords in the PodSecurity labeling scheme

liggitt commented on content/en/blog/_posts/2021-12-15-pod-security-admission-beta.md:
```diff
@@ -0,0 +1,774 @@
+   * `spec.securityContext`
+   * `spec.hostNetwork`
+   * `spec.volumes[*].hostPath`
+   * `spec.containers[*].securityContext`
```
the securityContext field itself isn't restricted, but specific subfields are, like `spec.containers[*].securityContext.privileged`

liggitt commented on content/en/blog/_posts/2021-12-15-pod-security-admission-beta.md:
```diff
@@ -0,0 +1,774 @@
+   * Restricted — Highly restricted, hardening against known and unknown privilege escalations. May cause compatibility issues
+
+Each of these policy levels define which fields are restricted within a pod specification and the allowed values. Some of the fields restricted by these policies include:
+   * `spec.securityContext`
```
the securityContext field itself isn't restricted, but specific subfields are, like `spec.securityContext.sysctls`

liggitt commented on content/en/blog/_posts/2021-12-15-pod-security-admission-beta.md:
```diff
@@ -0,0 +1,774 @@
+   * `spec.volumes[*].hostPath`
+   * `spec.containers[*].securityContext`
+
+Policy levels are applied via labels on Namespace resources, which allows for granular per-namespace policy selection. The AdmissionConfiguration in the api server can also be configured to set cluster-wide default levels and exemptions.
```
nit: api → API

liggitt commented on content/en/blog/_posts/2021-12-15-pod-security-admission-beta.md:
```diff
@@ -0,0 +1,774 @@
+
+### Policy modes
+
+Policies are applied in a specific mode. Multiple modes (with different policies) can be set on the same namespace. Here is a list of modes:
```
```suggestion
Policies are applied in a specific mode. Multiple modes (with different policy levels) can be set on the same namespace. Here is a list of modes:
```

liggitt commented on content/en/blog/_posts/2021-12-15-pod-security-admission-beta.md:
```diff
@@ -0,0 +1,774 @@
+   * `audit` — Violations will be recorded as an annotation in the audit logs, but don't affect whether the pod is allowed.
+   * `warn` — Violations will send a warning message back to the user, but don't affect whether the pod is allowed.
+
+In addition to modes you can also pin the policy to a specific version for example v1.22. Pinning to a specific version allows the behavior to remain consistent as the policy definition changes over Kubernetes releases. If pinning to a specific Pod Security Standard version in a Kubernetes release you should take the time to understand the Kubernetes [supported version policy](/releases/version-s ...
```
`... (for example, v1.22).`

liggitt replied:
`Pinning to a specific version allows the behavior to remain consistent if the policy definition changes in future Kubernetes releases.`

liggitt replied:
> ... you should take the time to understand the Kubernetes [supported version policy](/releases/version-skew-policy/#supported-versions).

I'm not sure that's particularly relevant... PodSecurity will continue to honor specific versions of the policy indefinitely.

lachie83 replied:
Thanks Jordan - the intent of this statement was if there was a specific bug in a Pod Security Standard that needed fixing and you were using a pinned version you may be out of luck depending on support. Not sure if that would help anyway because I'm not sure you can pin to a patch release (eg. v1.22.4)

lachie83 replied:
Yanking that sentence

liggitt commented on content/en/blog/_posts/2021-12-15-pod-security-admission-beta.md:
```diff
@@ -0,0 +1,774 @@
+
+### Applying a cluster-wide policy
+
+In addition to applying labels to namespaces to configure policy you can also configure cluster-wide policies and exemptions using the AdmissionConfiguration resource. 
+
+Using this resource, policy definitions are applied cluster-wide by default and any policy that is applied via namespace labels will take precedence. 
+
+There is no runtime configurable API for this resource and a cluster administrator would need to specify a path to the file below via the `--admission-control-config-file` flag on the API server. 
```
I wouldn't refer to the configuration type as a "resource"... that's typically reserved for URLs of REST APIs.

I'd probably call it "an `AdmissionConfiguration` configuration file."

liggitt commented on content/en/blog/_posts/2021-12-15-pod-security-admission-beta.md:
```diff
@@ -0,0 +1,774 @@
+
+In the following resource we are enforcing the baseline policy and warning and auditing the baseline policy. We are also making the kube-system namespace exempt from this policy.
+
+It's not recommended to alter control plane / clusters after install, so let's build a new cluster with a default policy on all namespaces.
```
on the other hand, this might be a chance to show up the safe rollout capabilities of this feature... you can tighten default warn/audit levels on an existing cluster without breaking writes to existing pods or workloads.

mehabhalodiya commented:
/assign @chrisnegus

sftim commented:
> It would also be ideal for both to be merged / published around the same dates so that they are available for consumption and most importantly the cross-links work

The easy answer is for us to have the tutorial page and the blog article both ready to go on v1.23 release day, and then manually merge both of these on a different day post release.

Slightly more faff, but possible: publish the tutorial at release time without a reference to the blog article, have the blog scheduled, have a follow-up PR ready that hyperlinks from the tutorial to the blog article as further reading.

sftim commented on content/en/blog/_posts/2021-12-15-pod-security-admission-beta.md:
```diff
@@ -0,0 +1,778 @@
+
+## How Does Pod Security Work?
+    
+Pod Security is a built-in [admission controller](/docs/reference/access-authn-authz/admission-controllers/) starting with Kubernetes v1.22, but can also be run as a standalone [webhook](/docs/reference/access-authn-authz/extensible-admission-controllers/). Admission controllers function by intercepting requests in the Kubernetes API server prior to persistence to storage. They can either `admit` ...
```
Is there a better place for “webhook” to hyperlink to / will there be, fairly soon?

liggitt replied:
https://github.com/kubernetes/website/blob/dev-1.23/content/en/docs/concepts/security/pod-security-admission.md#alternative-installing-the-podsecurity-admission-webhook is in the 1.23 branch (the anchor isn't very well-named though)

sftim commented on content/en/blog/_posts/2021-12-15-pod-security-admission-beta.md:
```diff
@@ -0,0 +1,778 @@
+
+### When to use `warn`?
+
+The typical uses for `warn` are:
```
```suggestion
The typical uses for `warn` are to get ready for a future change where you want to enforce
a different policy. The most two common cases would be:
```
How about case 3, you're not enforcing any policy yet, and you want to preview the impact? I think that's a good use for `warn` (maybe also: delete all existing Pods once you do).

I should add that I haven't actually tried Pod Security admission in any real-world cluster with a real world workload on it, so please take that in mind when considering my suggestions.

liggitt replied:
> How about case 3, you're not enforcing any policy yet, and you want to preview the impact? I think that's a good use for warn

That's sort of the same as "`warn` at a stricter level", where `enforce` is privileged and `warn` is baseline or restricted. If you think it's clearer to explicitly call that out as a third case, that's fine

> (maybe also: delete all existing Pods once you do).

If you're looking to check a new policy non-disruptively, deleting existing pods is not the way to do it :-)

sftim commented on content/en/blog/_posts/2021-12-15-pod-security-admission-beta.md:
```diff
@@ -0,0 +1,778 @@
+
+Each workload represents a higher level of security that would not pass the profile that comes after it.
+
+For the following examples, use the `busybox` container runs a `sleep` command for 1 million seconds (~11 days) or until deleted. Pod Security is not interested in the container, but rather the Kubernetes manifests and corresponding security values that are being admitted.
```
nits:
```suggestion
For the following examples, use the `busybox` container runs a `sleep` command for 1 million seconds (≅11 days) or until deleted. Pod Security is not interested in which container image you chose, but rather the Pod- and container- level settings and their implications for security. (There are third-party solutions that _do_ scan container images, and these also help with overall information security).
```

sftim commented on content/en/blog/_posts/2021-12-15-pod-security-admission-beta.md:
```diff
@@ -0,0 +1,778 @@
+
+**Authors:** Jim Angel (Google), Lachlan Evenson (Microsoft)
+
+With the release of Kubernetes v1.23, [Pod Security admission](/docs/concepts/security/pod-security-admission/) has now entered beta. Pod Security is a [built-in](/docs/reference/access-authn-authz/admission-controllers/) admission controller that evaluates pod specifications against a predefined set of [Pod Security Standards](/docs/concepts/security/pod-security-standards/) and determines wheth ...
```
```suggestion
With the release of Kubernetes v1.23, [Pod Security admission](/docs/concepts/security/pod-security-admission/) has now entered beta. Pod Security is a [built-in](/docs/reference/access-authn-authz/admission-controllers/) admission controller that evaluates pod specifications against a predefined set of [Pod Security Standards](/docs/concepts/security/pod-security-standards/) and determines whether to `admit` or `deny` the pod from running.

Pod Security is the successor to [PodSecurityPolicy](/docs/concepts/policy/pod-security-policy/) which was deprecated in the v1.21 release, and will be removed in Kubernetes v1.25. In this article, we will cover the key concepts of Pod Security along with how to use it. . We hope that cluster administrators and developers alike will use this new mechanism to enforce secure defaults for their workloads.
```

Could also hyperlink “specifications” to https://kubernetes.io/docs/concepts/overview/working-with-objects/kubernetes-objects/#object-spec-and-status if that sounds helpful.

sftim commented on content/en/blog/_posts/2021-12-15-pod-security-admission-beta.md:
```diff
@@ -0,0 +1,778 @@
+With the release of Kubernetes v1.23, [Pod Security admission](/docs/concepts/security/pod-security-admission/) has now entered beta. Pod Security is a [built-in](/docs/reference/access-authn-authz/admission-controllers/) admission controller that evaluates pod specifications against a predefined set of [Pod Security Standards](/docs/concepts/security/pod-security-standards/) and determines wheth ...
+
+## Why Pod Security
+
```
```suggestion

The overall aim of Pod Security is to let you isolate workloads. You can run a cluster that runs different workloads and, without adding extra third-party tooling, implement controls that require Pods for a workload to restrict their own privileges to a defined bounding set.

```

To me, the exact wording here isn't important, but a summary of what this new feature achieves is - separate from any mention of PodSecurityPolicy. Readers may never have encountered the PSP API, and may also not have read our previous blogs etc on the subject.

sftim reviewed (commented):
I think this could go in as-is, and the changes I'm suggesting here (are what I think would) add polish to an already good article.

sftim commented:
@jimangel I think this _is_ ready for review, right?

/approve
The intent of this PR is a change I am very happy to see made.

PushkarJ commented:
@sftim as per your feedback: https://github.com/kubernetes/website/pull/30502#issuecomment-973931518 (I think) the tutorial https://github.com/kubernetes/website/pull/30422/ is now ready to go and currently on `/hold` for this blog PR to get merged / published. 

Namespace level tutorial [preview](https://deploy-preview-30422--kubernetes-io-main-staging.netlify.app/docs/tutorials/security/ns-level-pss/) 
Cluster level tutorial [preview](https://deploy-preview-30422--kubernetes-io-main-staging.netlify.app/docs/tutorials/security/cluster-level-pss/)

karenhchu commented on content/en/blog/_posts/2021-12-15-pod-security-admission-beta.md:
```diff
@@ -0,0 +1,778 @@
+---
+layout: blog
+title: 'Pod Security Graduates to Beta'
+date: 2021-12-15
```
```suggestion
date: 2021-12-09
```

jimangel commented:
This is ready to merge / final review. I just pushed the following changes:

- Actionable feedback addressed
- TODO's removed
- `v1.23` links added
- Incorporated the future tutorial reference / links (thanks @PushkarJ!)
- Modified the file / date to match intended publish date
- General clean up / checks

At this point I don't anticipate making any further changes unless requested! Once the build finishes I'll update the preview links.

/cc @lachie83

sftim commented:
/hold cancel

v1.23 is released, I don't see anything blocking publication.

ritazh commented:
/approve
Thank you everyone for working on this PR! 🎉

sftim commented:
Not sure Prow saw the edited /lgtm from https://github.com/kubernetes/website/pull/30502#issuecomment-988419304 so:
/lgtm
